text/0000-impl-enum.md

```diff
@@ -0,0 +1,341 @@
+
+```rust
+struct Struct1;
+struct Struct2(...);
+
+match Struct2(...) {
+    Struct1 => ...,
+    ...
+}
+
+if let Struct2(...) = Struct1 {
+    ...
+}
+
+while let Struct2(...) = Struct1 {
+    ...
+}
+```
+
+None of these examples will work using normal structs, but variant structs work different.
+When two structs implement the same enum, only the matching branch stays. All other branches are removed at compile time.
+
+This already works, when not knowing the exact type, so writing generic functions, specialized on a single type, is easy.
+
+Whereever required, enums are implicitly cast to dynamic enums, for example when calling a function, that takes a dynamic enum. This is required to stay backwards compatible. Casting back to enum variants directly is not possible. Match should be used for that, just as before.
+
+## Enums in traits
+
+Since enums cannot be extended after definition, it's allowed to make traits, which contain generic functions, into objects, in case the generics are required to be enums:
+
+```rust
+trait Trait {
+    fn test<T: Enum>(&self, arg: Enum);
+}
+```
+
+The vtable of this trait will contain a new entry for every variant of the enum type. In case of multiple enums as arguments, it will just generate even more entries. Using multiple dynamic enums for traits, that will be made into an object, should be avoided.
+
+It may be useful to define a trait, specialized for multiple different enums. This will look like this:
+
+```rust
+trait Trait {
+    enum Enum;
+    fn test<T: Self::Enum>(&self, arg: T);
+}
+```
+
+## Example
+
+
+Using these new feature, writing an efficient version of the motivational example is now possible:
+
+```rust
+use recieve_exit_event; // fn() -> bool
+
+// remove generic type from 
+trait Updateable {
+    // add associated enum
+    enum Event;
+    // require associated enum
+    fn update<E: Event>(&self, event: E) -> bool; 
+}
+
+enum Event {
+    Idle,
+    Exit,
+}
+
+struct Object;
+impl Updateable for Object {
+    // update implementation to match new version of Update
+    enum Event = Event;
+    fn update<E: Event>(&self, event: E) -> bool {
+        match event {
+            Event::Idle => true,
+            Event::Exit => false,
+        }
+    }
+}
+
+fn run() {
+    let current_updatable = (&Object as &dyn Updateable<Event = Event>);
+    loop {
+        if recieve_exit_event() {
+            if !current_updatable.update(Event::Exit) {
+                break;
+            }
+        }
+        if !current_updatable.update(Event::Idle) {
+            break;
+        }
+    }
+}
+```
+
+As it's easy to see, with this feature, there are almost no differences in the implementation, but the overhead from using enums is eliminated.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+It seems the explanation in the previous part is already detailed enough for most parts and moving parts down here will just make it less clear.
```

Comment:
Not at all. It's not even clear at a high level from the guide what this does let alone what impact it would have on Rust's 
static, dynamic semantics, and syntax.